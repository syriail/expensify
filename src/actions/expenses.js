import uuid from 'uuid';
import database from '../firebase/firebase';

export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {})=>{
    return (dispatch)=>{
        const {
            description = '', note = '', amount = 0, createAt = 0
        } = expenseData;

        const expense = {description, note, amount, createAt};
        console.log(expense);
        //We return here to chain another then in test cases. See expenses.test.js
        return database.ref('expenses').push(expense).then((ref)=>{
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};

export const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

