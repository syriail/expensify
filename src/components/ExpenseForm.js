import React from 'react';
import {connect} from 'react-redux';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';

const now = moment();


export default class ExpenseForm extends React.Component{

    constructor(props){
        super(props);
          this.state = {
            description: props.expense ? props.expense.description : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            note: props.expense ? props.expense.note : '',
            createAt: props.expense ? moment(props.expense.createAt) : moment(),
            calendarFocused: false,
            error: ''
        };
    }
    /* User this way when we don't need to initialize state depending on props
    
    state = {
        description: '',
        amount: '',
        note: '',
        createAt: moment(),
        calendarFocused: false,
        error: ''
    };
    */

    onDescriptionChange = (e)=>{
        const description = e.target.value;
        this.setState(() => ({description}));
    };

    onNoteChange = (e)=>{
        const note = e.target.value;
        this.setState(()=>({note}));
    };

    onAmountChange = (e)=>{
        const amount = e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(()=>({amount}));
        }
    };

    onDateChange = (createAt)=>{
        if(createAt){
            this.setState(()=>({createAt}));
        }
        
    };

    onFocusChange = ({focused})=>{

        this.setState(()=>({calendarFocused: focused}));
    };

    onSubmit = (e)=>{
        e.preventDefault();
        const description = this.state.description;
        const amount = this.state.amount;
        if(!this.state.description || !this.state.amount){
            //Show the error: Please provide a description and an amount
            this.setState(()=>({error: 'Please provide a description and an amount'}));
        }else{
            this.setState(()=>({error: ''}));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10)*100,
                note: this.state.note,
                createAt: this.state.createAt.valueOf()
            });
        }
    };

    render(){
        return (
            <div>
            {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="text"
                        placeholder="amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={()=>false}
                    />
                    <textarea
                        placeholder="Enter here an optional note for your expense"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    />
                    <button>Save</button>
                </form>
            </div>
        );
    }
}
