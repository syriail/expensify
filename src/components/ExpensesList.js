import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpensesList = (props) =>(
    <div>
        <h1> Expenses List</h1>
        {
            props.expenses.length === 0? (
                <p>No expenses have been added</p>
            ) : (
                props.expenses.map((expense)=>{
                    return <ExpenseListItem key={expense.id} {...expense}/>
                })
            )
        }
        
    </div>
);

const mapStateToProps = (state) =>{
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpensesList);
