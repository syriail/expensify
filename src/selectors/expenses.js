import moment from 'moment';

export default (expenses, {text, sortBy, startDate, endDate})=>{
    var visibleExpenses = expenses.filter(({description, createAt}) =>{
        const createAtMoment = moment(createAt);
        const textMatch = description.toLowerCase().includes(text.toLowerCase());
        const startDateMatch = startDate ? startDate.isSameOrBefore(createAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createAtMoment, 'day') : true;
        return textMatch && startDateMatch && endDateMatch;
    }).sort((a, b)=>{
        if(sortBy === 'date'){
            return a.createAt < b.createAt ? 1 : -1;
        }else{
            return a.amount < b.amount ? 1 : -1;
        }
    });
    return visibleExpenses;
};