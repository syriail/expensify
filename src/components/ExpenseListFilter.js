import React from 'react';
import {connect} from 'react-redux'
import {setEndDate, setStartDate, setTextFilter, sortByAmount, sortByDate} from '../actions/filters';
import {DateRangePicker} from 'react-dates';

export class ExpenseListFilter extends React.Component{
    state = {
        calendarFocused: null
    };

    onDatesChange = ({startDate, endDate})=>{
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (calendarFocused)=>{
        this.setState(()=>({calendarFocused}));
    };

    onTextChange = (e)=>{
        this.props.setTextFilter(e.target.value);
    };

    onSortChange = (e)=>{
        if(e.target.value === 'amount'){
            this.props.sortByAmount();
        }else{
            this.props.sortByDate();
        }
    };
    render() {
        return (
        <div>
            <input 
                type="text" 
                value={this.props.filters.text}
                onChange={this.onTextChange}
            >
            </input>
            <select onChange={this.onSortChange}>
                <option value="amount">Amount</option>    
                <option value="date">Date</option>
                
            </select>

            <DateRangePicker 
                startDate={this.props.filters.startDate}
                endDate={this.props.filters.endDate}
                onDatesChange={this.onDatesChange}
                focusedInput={this.state.calendarFocused}
                onFocusChange={this.onFocusChange}
                showClearDates={true}
                numberOfMonths={1}
                isOutsideRange={()=>false}
            />

        </div>
        );
    };
}

const mapStateToProps = (state) =>{
    return {
        filters: state.filters
    }
};

const mapDispatchToProps = (dispatch)=>{
    return {
        setTextFilter: (text)=>{dispatch(setTextFilter(text))},
        sortByAmount: ()=>{dispatch(sortByAmount())},
        sortByDate: ()=> {dispatch(sortByDate())},
        setStartDate: (startDate)=>{dispatch(setStartDate(startDate))},
        setEndDate: (endDate)=>{dispatch(setEndDate(endDate))}

    };
};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);