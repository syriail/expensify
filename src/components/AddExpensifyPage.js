import React from 'react';
import ReactDOM from 'react-dom';
import ExpenseForm from './ExpenseForm';
import {connect} from 'react-redux';
import {addExpense} from '../actions/expenses';

export class AddExpensifyPage extends React.Component{
    onSubmit = (expense)=>{
        //props.dispatch(addExpense(expense));
                //The solution is easier to test
        this.props.addExpense(expense);
        this.props.history.push('/');
    };
    render(){
        return (
            <div>
            Add Expensify page
            <ExpenseForm 
                onSubmit={this.onSubmit}
            />
            </div>
        );
        
    };
}

const mapDispatchToProps = (dispatch)=>{
    return {
        addExpense: (expense)=> dispatch(addExpense(expense))
    };
};

export default connect(undefined, mapDispatchToProps)(AddExpensifyPage);