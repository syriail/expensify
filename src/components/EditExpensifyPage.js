import React from 'react';
import ReactDOM from 'react-dom';
import {connect} from 'react-redux';
import ExpenseForm from '../components/ExpenseForm';
import {startEditExpense, startRemoveExpense} from '../actions/expenses';

export class EditExpensifyPage extends React.Component{

    onSubmit = (expense)=>{
        //props.dispatch(startEditExpense(props.expense.id, {...expense}));
        //The following aproach is better for testing
        this.props.startEditExpense(this.props.expense.id, {...expense});
        this.props.history.push('/');
    };

    onClick = ()=>{
        //props.dispatch(removeExpense({id: props.expense.id}));
        this.props.startRemoveExpense(this.props.expense.id);
        this.props.history.push('/');
    };

    render(){
        return (
            <div>
                <ExpenseForm 
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button
                onClick={this.onClick}
            >Remove
            </button>
            </div>
        );
    };
}

const mapStateToProps = (state, props)=>{
    return {
        expense: state.expenses.find((expense)=>{
            return expense.id === props.match.params.id;
        })
    }
}

const mapDispatchToProps = (dispatch)=>{
    return {
        startEditExpense: (id, expense)=> dispatch(startEditExpense(id, expense)),
        startRemoveExpense: (id)=> dispatch(startRemoveExpense({id}))
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensifyPage);