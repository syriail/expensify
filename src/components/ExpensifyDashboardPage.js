import React from 'react';
import ReactDOM from 'react-dom';
import ExpensesList from './ExpensesList';
import ExpensesListFilter from './ExpenseListFilter';
import ExpenseListFilter from './ExpenseListFilter';
import ExpensesSummary from './ExpensesSummary';

const ExpensifyDashboardPage = () => (
    <div>
        <ExpenseListFilter />
        <ExpensesList />
        <ExpensesSummary />
    </div>
);

export default ExpensifyDashboardPage;