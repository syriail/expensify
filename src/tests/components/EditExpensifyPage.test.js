import React from 'react';
import {shallow} from 'enzyme';
import {EditExpensifyPage} from '../../components/EditExpensifyPage';
import expenses from '../fixtures/expenses';

let editExpenseSpy, removeExpenseSpy, history, wrapper;

beforeEach(()=>{
    editExpenseSpy = jest.fn();
    removeExpenseSpy = jest.fn();
    history = {push: jest.fn()};
    wrapper = shallow(<EditExpensifyPage
        editExpense={editExpenseSpy}
        removeExpense={removeExpenseSpy}
        history={history}
        expense={expenses[1]}
        />);
});

test('Should render edit page correctly', ()=>{
    expect(wrapper).toMatchSnapshot();
});

test('Should submit form after editing expense', ()=>{
    const expense = expenses[1];
    wrapper.find('ExpenseForm').prop('onSubmit')(expense);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(editExpenseSpy).toHaveBeenLastCalledWith(expense.id, expense);
});

test('Should remove expense', ()=>{
    const expense = expenses[1];
    wrapper.find('button').simulate('click');
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(removeExpenseSpy).toHaveBeenLastCalledWith(expense.id);
});