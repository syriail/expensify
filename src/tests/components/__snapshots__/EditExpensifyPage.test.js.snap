// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render edit page correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 350,
        "createAt": 172800000,
        "description": "Tee",
        "id": 2,
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
